hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/ChecksumFs.java

```diff
@@ -457,17 +458,35 @@ public boolean setReplication(Path src, short replication)
   @Override
   public void renameInternal(Path src, Path dst) 
     throws IOException, UnresolvedLinkException {
+    renameInternal(src, dst, (s, d) -> getMyFs().rename(s, d));
+  }
+
+  @Override
+  public void renameInternal(Path src, Path dst, boolean overwrite)
+      throws AccessControlException, FileAlreadyExistsException,
+      FileNotFoundException, ParentNotDirectoryException,
+      UnresolvedLinkException, IOException {
+    Options.Rename renameOpt = Options.Rename.NONE;
+    if (overwrite) {
+      renameOpt = Options.Rename.OVERWRITE;
+    }
+    final Options.Rename opt = renameOpt;
+    renameInternal(src, dst, (s, d) -> getMyFs().rename(s, d, opt));
```

Comment:
Would it be wrong to delegate to myFs like here: https://github.com/apache/hadoop/blob/d992c956a421cc5731b80d74355703360bcab648/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/FilterFs.java#L254

Reply:
ChecksumFs should deal with checksum file by itself: that's why the method should be override. Just delegating rename would have same result, checksum file leak.

> What is the concrete class of the AbastractFileSystem field?

If the source file is from local filesystem, I think it's RawLocalFs. It could be DelegateToFileSystem, but it will redirect to RawLocalFs eventually.

Reply:
Looking at this, its the more elegant functional API. Which is nice for hadoop 3+.  But I fear its probably going to lose all that elegance on branch-2 (assuming you do want a backport). If you do, then simply copying the existing renameInternal to one with a new signature is going to be the simplest